subprojects/usage-graph-generator/src/main/kotlin/org/cafejojo/schaapi/usagegraphgenerator/InstructionGrouper.kt

```
@@ -0,0 +1,128 @@
+
+        cfg.method.instructions.iterator().forEach(fun(instruction: AbstractInsnNode) {
+            if (!block.containsInstructionNum(instructionIndex++)) return
+
+            if (instruction is LabelNode) {
+                val current = if (instruction == lastLabel && isBranchBlock(block)) BranchNode() else StatementNode()
+                last?.successors?.add(current)
+                last = current
+                if (first == null) first = last
+            }
+
+            if (instruction.opcode < 0 || instruction.opcode >= Printer.OPCODES.size) return
+
+            if (last == null) throw IllegalStateException("Instruction found before label")
+
+            last?.instructions?.add(instruction)
+        })
+
+        return Pair(first, last)
+    }
+
+    /**
+     * Finds the last label within the block.
+     *
+     * @param block CFG block.
+     * @return last label within the given block.
+     */
+    private fun findLastLabelInBlock(block: IBlock): LabelNode? {
+        var instructionIndex = 0
+        var labelNode: LabelNode? = null
+        cfg.method.instructions.iterator().forEach(fun(instruction) {
+            if (!block.containsInstructionNum(instructionIndex++)) return
+            if (instruction is LabelNode) labelNode = instruction
+        })
+        return labelNode
+    }
+
+    /**
+     * Determines if the [block] under evaluation is relevant.
+     *
+     * @param block CFG block.
+     * @return boolean indicating if the [block] under evaluation is relevant.
+     */
+    private fun isProcessableBlock(block: IBlock) =
+        block is MethodEntry || block is MethodExit || !block.isVirtual(cfg.method)
+
+    /**
+     * Determines if the given [block] is a branching block.
+     *
+     * @param block CFG block.
+     * @return boolean indicating if the given [block] is a branching block.
+     */
+    private fun isBranchBlock(block: IBlock) = block.edges.size == 3 && block.edges.count { it.src == block } == 2
```

Is it possible to have blocks with more than three edges?

---

Probably not, not sure at this time. Will find this out in a later stage.

---

I'd throw an exception if that happens, or at least place an assert. Makes it easier to detect whether our current assumptions hold.

---

Removed the `block.edges.size == 3`, since it's actually not relevant